# Solutions to Lesson 4 Exercises

For each exercise, the solutions below show one possible way of solving it, but you might have used a different approach, and that's great! There is almost always more than one way to solve any particular problem in Python.

## Initial Setup

Since this notebook is in the `solutions` sub-folder, use the  magic command `%cd` to go up one folder to the main project folder to keep the file paths the same as in the lessons:

In [1]:
%cd ..

C:\Users\jenfl\Projects\eoas-python


Import libraries, display plots inline, initialize variables from lesson:

In [2]:
import pandas
%matplotlib inline

# Import library from local .py file:
import ecweather

In [3]:
world = pandas.read_csv('data/gapminder_world_data_2018.csv')
world['area'] = world['population'] / world['pop_density']

# Exercise 4

Working with our DataFrame `world` containing the data from `data/gapminder_world_data_2018.csv`:

**a)** Sort `world` on the `'pop_density'` column from lowest to highest and display the first 10 rows.
- Does Canada appear anywhere in the results? Which country has the lowest population density?

In [4]:
world.sort_values('pop_density').head(10)

country  population    region                       sub_region  \
105    Mongolia     3120000      Asia                     Eastern Asia   
110     Namibia     2590000    Africa               Sub-Saharan Africa   
7     Australia    24800000   Oceania        Australia and New Zealand   
70      Iceland      338000    Europe                  Northern Europe   
150    Suriname      568000  Americas  Latin America and the Caribbean   
91        Libya     6470000    Africa                  Northern Africa   
66       Guyana      782000  Americas  Latin America and the Caribbean   
28       Canada    37000000  Americas                 Northern America   
21     Botswana     2330000    Africa               Sub-Saharan Africa   
101  Mauritania     4540000    Africa               Sub-Saharan Africa   

     income_group  life_expectancy  gdp_per_capita  children_per_woman  \
105  Lower middle             68.7           12100                2.67   
110  Upper middle             65.7           10500                3.30   
7            High             82.9           45800                1.83   
70           High             82.6           47400                1.91   
150  Upper middle             71.6           13200                2.32   
91   Upper middle             75.5           18300                2.20   
66   Upper middle             68.2            7740                2.46   
28           High             82.2           43800                1.56   
21   Upper middle             66.9           16500                2.64   
101  Lower middle             70.6            3670                4.55   

     child_mortality  pop_density          area  
105            16.80         2.01  1.552239e+06  
110            43.40         3.14  8.248408e+05  
7               3.40         3.22  7.701863e+06  
70              2.08         3.37  1.002967e+05  
150            19.20         3.64  1.560440e+05  
91             12.20         3.68  1.758152e+06  
66             31.80         3.97  1.969773e+05  
28              4.82         4.06  9.113300e+06  
21             36.60         4.12  5.655340e+05  
101            80.60         4.40  1.031818e+06

- Yes, Canada is one of the 10 most sparsely populated countries.
- Mongolia has the lowest population density

- Now sort from highest to lowest and display the first 10 rows. Which country has the highest population density?

In [5]:
world.sort_values('pop_density', ascending=False).head(10)

country  population    region                       sub_region  \
139    Singapore     5790000      Asia               South-eastern Asia   
11       Bahrain     1570000      Asia                     Western Asia   
98      Maldives      444000      Asia                    Southern Asia   
100        Malta      432000    Europe                  Southern Europe   
12    Bangladesh   166000000      Asia                    Southern Asia   
121    Palestine     5050000      Asia                     Western Asia   
13      Barbados      286000  Americas  Latin America and the Caribbean   
102    Mauritius     1270000    Africa               Sub-Saharan Africa   
88       Lebanon     6090000      Asia                     Western Asia   
145  South Korea    51200000      Asia                     Eastern Asia   

     income_group  life_expectancy  gdp_per_capita  children_per_woman  \
139          High             84.0           83900                1.26   
11           High             77.2           44300                1.99   
98   Upper middle             80.1           13000                2.01   
100          High             81.8           37900                1.48   
12   Lower middle             73.4            3720                2.05   
121  Lower middle             72.2            2700                3.88   
13           High             76.8           16000                1.80   
102  Upper middle             74.9           21000                1.43   
88   Upper middle             80.5           13400                1.71   
145          High             81.3           36800                1.33   

     child_mortality  pop_density           area  
139             2.52       8270.0     700.120919  
11              7.10       2060.0     762.135922  
98              7.75       1480.0     300.000000  
100             6.32       1350.0     320.000000  
12             32.00       1280.0  129687.500000  
121            18.60        839.0    6019.070322  
13             11.50        666.0     429.429429  
102            13.20        625.0    2032.000000  
88              7.97        596.0   10218.120805  
145             3.08        526.0   97338.403042

Singapore has the highest population density (8270 people per square kilometre!)

**b)** Let's find out how many countries have life expectancy greater than 80 years and the total number of people living in those countries.
- Create a Boolean Series `long_life` which is `True` for rows where `world['life_expectancy']` is greater than 80, and `False` otherwise.
- Take the sum of `long_life` to find out how many countries have life expectancy > 80.
- Use `long_life` as a filter to extract the rows of `world` with life expectancy > 80, and store the output as a new DataFrame `world_long_life`.
- Use the `sum` method on the `population` column of `world_long_life` to find the total population in these countries. 

In [6]:
long_life = world['life_expectancy'] > 80
long_life.sum()

34

34 of the countries have life expectancy above 80 years.

In [7]:
world_long_life = world[long_life]
world_long_life.head()

country  population    region                       sub_region  \
7   Australia    24800000   Oceania        Australia and New Zealand   
8     Austria     8750000    Europe                   Western Europe   
15    Belgium    11500000    Europe                   Western Europe   
28     Canada    37000000  Americas                 Northern America   
31      Chile    18200000  Americas  Latin America and the Caribbean   

   income_group  life_expectancy  gdp_per_capita  children_per_woman  \
7          High             82.9           45800                1.83   
8          High             81.8           44600                1.52   
15         High             81.2           42800                1.80   
28         High             82.2           43800                1.56   
31         High             80.7           23400                1.76   

    child_mortality  pop_density          area  
7              3.40         3.22  7.701863e+06  
8              2.89       106.00  8.254717e+04  
15             3.55       380.00  3.026316e+04  
28             4.82         4.06  9.113300e+06  
31             7.62        24.50  7.428571e+05

In [8]:
world_long_life['population'].sum()

747104000

The total population in these 34 countries is about 747 million.

**c)** Find the total population in each income group by grouping `world` on the `'income_group'` column and taking the sum of the `'population'` column.

In [9]:
world.groupby('income_group')['population'].sum()

income_group
High            1212463200
Low              752002000
Lower middle    3006961000
Upper middle    2622182000
Name: population, dtype: int64

### Bonus exercises

**d)** Repeat (a) for the `'population'` column to find the countries with the lowest and highest populations.

In [10]:
world.sort_values('population').head(10)

country  population    region  \
137           Seychelles       95200    Africa   
4    Antigua and Barbuda      103000  Americas   
62               Grenada      108000  Americas   
160                Tonga      109000   Oceania   
83              Kiribati      118000   Oceania   
133                Samoa      198000   Oceania   
172              Vanuatu      282000   Oceania   
13              Barbados      286000  Americas   
70               Iceland      338000    Europe   
16                Belize      382000  Americas   

                          sub_region  income_group  life_expectancy  \
137               Sub-Saharan Africa          High             74.2   
4    Latin America and the Caribbean          High             77.6   
62   Latin America and the Caribbean  Upper middle             71.9   
160                        Polynesia  Upper middle             70.7   
83                        Micronesia  Lower middle             62.2   
133                        Polynesia  Upper middle             72.4   
172                        Melanesia  Lower middle             64.3   
13   Latin America and the Caribbean          High             76.8   
70                   Northern Europe          High             82.6   
16   Latin America and the Caribbean  Upper middle             72.5   

     gdp_per_capita  children_per_woman  child_mortality  pop_density  \
137           27500                2.25            12.70       207.00   
4             21000                2.03             7.89       234.00   
62            13500                2.06            14.60       319.00   
160            5500                3.56            16.10       151.00   
83             1890                3.57            52.50       146.00   
133            5950                3.88            16.70        69.90   
172            2900                3.20            26.00        23.10   
13            16000                1.80            11.50       666.00   
70            47400                1.91             2.08         3.37   
16             7710                2.44            13.80        16.80   

              area  
137     459.903382  
4       440.170940  
62      338.557994  
160     721.854305  
83      808.219178  
133    2832.618026  
172   12207.792208  
13      429.429429  
70   100296.735905  
16    22738.095238

The island nation of Seychelles has the lowest population.

In [11]:
world.sort_values('population', ascending=False).head(10)

country  population    region                       sub_region  \
32           China  1420000000      Asia                     Eastern Asia   
71           India  1350000000      Asia                    Southern Asia   
169  United States   327000000  Americas                 Northern America   
72       Indonesia   267000000      Asia               South-eastern Asia   
22          Brazil   211000000  Americas  Latin America and the Caribbean   
120       Pakistan   201000000      Asia                    Southern Asia   
116        Nigeria   196000000    Africa               Sub-Saharan Africa   
12      Bangladesh   166000000      Asia                    Southern Asia   
131         Russia   144000000    Europe                   Eastern Europe   
103         Mexico   131000000  Americas  Latin America and the Caribbean   

     income_group  life_expectancy  gdp_per_capita  children_per_woman  \
32   Upper middle             76.9           16000                1.64   
71   Lower middle             69.1            6890                2.28   
169          High             79.1           54900                1.90   
72   Lower middle             72.0           11700                2.31   
22   Upper middle             75.7           14300                1.70   
120  Lower middle             68.0            5220                3.35   
116  Lower middle             66.1            5570                5.39   
12   Lower middle             73.4            3720                2.05   
131  Upper middle             71.1           24800                1.76   
103  Upper middle             76.8           17300                2.12   

     child_mortality  pop_density          area  
32              9.95       151.00  9.403974e+06  
71             41.10       455.00  2.967033e+06  
169             6.06        35.70  9.159664e+06  
72             25.00       147.00  1.816327e+06  
22             14.20        25.20  8.373016e+06  
120            76.80       260.00  7.730769e+05  
116            97.90       215.00  9.116279e+05  
12             32.00      1280.00  1.296875e+05  
131             7.33         8.79  1.638225e+07  
103            13.80        67.30  1.946508e+06

China has the highest population.

**e)** Building on your results from (b), what *fraction* of the world's total population lives in countries with life expectancy above 80 years?

In [12]:
world_long_life['population'].sum() / world['population'].sum()

0.09838590302828634

**f)** How many countries have life expectancy > 80 *and* are in the "Upper middle" income group? Which countries are they? Hints:
- In Lesson 1, we saw the `and`, `or`, and `not` logical operators. When working with arrays of data such `pandas` Series and DataFrames, we have to use different operators: `&` (and), `|` (or), `~` (not). So for example, to find which countries in Europe have population above 20 million, we could create the Boolean Series:

```python
europe_high_pop = (world['region'] == 'Europe') & (world['population'] > 2e7)
```

- Note the parentheses around the expressions on either side of `&` - these are very important! The `&` operator has precedence over the `==` and `>` comparison operators, so parentheses are necessary to ensure the correct order of operations.

In [13]:
long_life_uppermid = (world['life_expectancy'] > 80) & (world['income_group'] == 'Upper middle')
long_life_uppermid.sum()

4

In [14]:
world[long_life_uppermid]

country  population    region                       sub_region  \
37   Costa Rica     4950000  Americas  Latin America and the Caribbean   
88      Lebanon     6090000      Asia                     Western Asia   
98     Maldives      444000      Asia                    Southern Asia   
125        Peru    32600000  Americas  Latin America and the Caribbean   

     income_group  life_expectancy  gdp_per_capita  children_per_woman  \
37   Upper middle             81.4           16200                1.75   
88   Upper middle             80.5           13400                1.71   
98   Upper middle             80.1           13000                2.01   
125  Upper middle             80.2           12700                2.34   

     child_mortality  pop_density          area  
37              8.71         97.0  5.103093e+04  
88              7.97        596.0  1.021812e+04  
98              7.75       1480.0  3.000000e+02  
125            14.30         25.4  1.283465e+06

There are 4 countries matching this criteria: Costa Rica, Lebanon, Maldives, and Peru.

**g) Data wrangling - downloading and concatenating Environment Canada weather data**

Now we'll put all the pieces together to download and concatenate YVR weather data from Environment Canada!
- First, I recommend making a copy of the `data/raw` sub-folder so that you can preserve the original files in case you need them.
- If you haven't already, complete previous bonus exercise 2(f) to import the local file `ecweather.py` as a library.

Copy the code below into your notebook and run it:

```python
# Environment Canada station ID code for Vancouver Airport
station_id = 889

# Save CSV file for 1938
year = 1938
savefile = 'data/raw/downloaded_weather_YVR_' + str(year) + '.csv'
ecweather.download_daily_raw(env_canada_id=station_id, year=year, savefile=savefile)
```

If it ran successfully, you'll have a new file `downloaded_weather_YVR_1938.csv` in the `data/raw` sub-folder. Check out the file in JupyterLab CSV viewer.

In [15]:
# Environment Canada station ID code for Vancouver Airport
station_id = 889

# Save CSV file for 1938
year = 1938
savefile = 'data/raw/downloaded_weather_YVR_' + str(year) + '.csv'
ecweather.download_daily_raw(env_canada_id=station_id, year=year, savefile=savefile)

Saving to data/raw/downloaded_weather_YVR_1938.csv


Next, using the code above and previous bonus exercise 3(h) as a guide, create a loop which loops over the years 1938-2012 and downloads the CSV file for each year. Run your code and check the files in the `data/raw` sub-folder to make sure the files were saved properly.

> Note: Environment Canada changed its station ID code for Vancouver Airport in mid-2013, which completes things. To keep this exercise simple, we'll only work with data from 1938-2012.

In [16]:
for year in range(1938, 2013):
    savefile = 'data/raw/downloaded_weather_YVR_' + str(year) + '.csv'
    ecweather.download_daily_raw(env_canada_id=station_id, year=year, savefile=savefile)

Saving to data/raw/downloaded_weather_YVR_1938.csv
Saving to data/raw/downloaded_weather_YVR_1939.csv
Saving to data/raw/downloaded_weather_YVR_1940.csv
Saving to data/raw/downloaded_weather_YVR_1941.csv
Saving to data/raw/downloaded_weather_YVR_1942.csv
Saving to data/raw/downloaded_weather_YVR_1943.csv
Saving to data/raw/downloaded_weather_YVR_1944.csv
Saving to data/raw/downloaded_weather_YVR_1945.csv
Saving to data/raw/downloaded_weather_YVR_1946.csv
Saving to data/raw/downloaded_weather_YVR_1947.csv
Saving to data/raw/downloaded_weather_YVR_1948.csv
Saving to data/raw/downloaded_weather_YVR_1949.csv
Saving to data/raw/downloaded_weather_YVR_1950.csv
Saving to data/raw/downloaded_weather_YVR_1951.csv
Saving to data/raw/downloaded_weather_YVR_1952.csv
Saving to data/raw/downloaded_weather_YVR_1953.csv
Saving to data/raw/downloaded_weather_YVR_1954.csv
Saving to data/raw/downloaded_weather_YVR_1955.csv
Saving to data/raw/downloaded_weather_YVR_1956.csv
Saving to data/raw/downloaded_w

Finally, create a for loop, based on your work in previous bonus exercise 3(h), to read each CSV file into a DataFrame and concatenate them all together into a single DataFrame. Check that the shape, dates, etc. in the merged data make sense, and save the merged data to a CSV file `data/merged_weather_YVR_1938-2012.csv`.

**Data wrangling mission completed! Great work!!!**

> If you want to see all the nitty gritty details that went into creating the full 1938-2017 file we used in earlier exercises, check out the notebooks in the `wrangling-env-canada` folder. 

In [17]:
data_merged = pandas.DataFrame()

for year in range(1938, 2013):
    data_file = 'data/raw/downloaded_weather_YVR_' + str(year) + '.csv'
    data_in = pandas.read_csv(data_file, skiprows=24)
    data_merged = data_merged.append(data_in, ignore_index=True)

In [18]:
data_merged.shape

(27394, 27)

In [19]:
data_merged.head(2)

Date/Time  Year  Month  Day  Data Quality  Max Temp (°C) Max Temp Flag  \
0  1938-01-01  1938      1    1           NaN            9.4           NaN   
1  1938-01-02  1938      1    2           NaN            7.2           NaN   

   Min Temp (°C) Min Temp Flag  Mean Temp (°C)          ...           \
0           -0.6           NaN             4.4          ...            
1            1.7           NaN             4.5          ...            

  Total Snow (cm)  Total Snow Flag Total Precip (mm)  Total Precip Flag  \
0             NaN                M               0.3                NaN   
1             NaN                M               0.5                NaN   

  Snow on Grnd (cm)  Snow on Grnd Flag Dir of Max Gust (10s deg)  \
0               NaN                NaN                       NaN   
1               NaN                NaN                       NaN   

   Dir of Max Gust Flag Spd of Max Gust (km/h)  Spd of Max Gust Flag  
0                   NaN                    NaN                   NaN  
1                   NaN                    NaN                   NaN  

[2 rows x 27 columns]

In [20]:
data_merged.tail(2)

Date/Time  Year  Month  Day  Data Quality  Max Temp (°C)  \
27392  2012-12-30  2012     12   30           NaN            5.0   
27393  2012-12-31  2012     12   31           NaN            4.0   

      Max Temp Flag  Min Temp (°C) Min Temp Flag  Mean Temp (°C)  \
27392           NaN           -1.0           NaN             2.0   
27393           NaN           -0.4           NaN             1.8   

               ...          Total Snow (cm)  Total Snow Flag  \
27392          ...                      0.0              NaN   
27393          ...                      0.0                T   

      Total Precip (mm)  Total Precip Flag Snow on Grnd (cm)  \
27392               0.0                NaN               0.0   
27393               1.2                NaN               0.0   

       Snow on Grnd Flag Dir of Max Gust (10s deg)  Dir of Max Gust Flag  \
27392                NaN                       NaN                   NaN   
27393                NaN                       NaN                   NaN   

      Spd of Max Gust (km/h)  Spd of Max Gust Flag  
27392                    <31                   NaN  
27393                    <31                   NaN  

[2 rows x 27 columns]

In [21]:
data_merged['Year'].unique()

array([1938, 1939, 1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947, 1948,
       1949, 1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959,
       1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970,
       1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981,
       1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992,
       1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003,
       2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012], dtype=int64)

In [22]:
data_merged.to_csv('data/merged_weather_YVR_1938-2012.csv', index=False)

---

Go to: [solutions-1](solutions-1.ipynb) | [solutions-2](solutions-2.ipynb) | [solutions-3](solutions-3.ipynb) | solutions-4 | [solutions-5](solutions-5.ipynb) | [solutions-6](solutions-6.ipynb)

[home](../0-introduction.ipynb)